doc/neps/nep-0023-restructuring-np-core.rst

```diff
@@ -0,0 +1,144 @@
+==============================================
+NEP 23 - Making details of ``np.core`` private
+==============================================
+
+:Author: Eric Wieser <wieser.eric+numpy@gmail.com>
+:Status: Draft
+:Type: Standards Track
+:Created: 2018-09-09
+:Resolution: <url> (required for Accepted | Rejected | Withdrawn)
+
+Abstract
+--------
+
+The ``np.core`` package consists of a collection of submodules, most of which have a name not prefixed with an underscore - the pythonic way to indicate a public API.
+As a result, the full set of members within any of these modules can be considered public API.
+Since these submodules import names from other submodules, these imported names appear to be part of the public API too.
+
+As a result of this, any of the following can be seen as a breaking change:
+
+* Moving a function from one file to another
+* Removing an "unused" import from a file
+
+This traps us into creating increasingly convoluted [1]_ and cyclic imports, under the constant fear that a reorganization will break downstream code.
+
+This NEP does not consider ``np.lib``.
+
+Detailed description
+--------------------
+
+The list of existing modules within ``np.core`` is::
+
+	>>> import inspect
+	>>> [name for name, mod in inspect.getmembers(np.core, inspect.ismodule)]
+	['_internal',
+	 '_methods',
+	 'arrayprint',
+	 'char',
+	 'defchararray',
+	 'einsumfunc',
+	 'fromnumeric',
+	 'function_base',
+	 'getlimits',
+	 'info',
+	 'machar',
+	 'multiarray',
+	 'numeric',
+	 'numerictypes',
+	 'rec',
+	 'records',
+	 'shape_base',
+	 'umath']
+
+The publicness of this API surface is contagious - in principle, ``np.core.numerictypes.array`` (an alias of ``np.core.multiarray.array``) is part of the API too.
+
+This is problematic - it means that if ``numeric.py`` contains ``from numpy.core.multiarray import array``, then we can never remove that line, without risking breaking someone using ``np.core.numeric.array``.
+
+The proposal here is to:
+
+* Make all of the existing modules within `np.core` private, adding a leading underscore to their names.
```

Comment:
Maybe a lot less work: move `np.core` to `np._core` (and and import warnings for `np.core`)

Reply:
@rgommers A similar thought occurred to me. This is beginning to sound like the evolution of the `numpy.testing` package where we ended up with a couple of stub modules and `_private`. A clear statement of intent and justification would be good to have.

Reply:
I see that there is a statement of intent. Maybe we can expand this to a discussion of what we would like NumPy to look like in the long term.